README.md

```diff
@@ -8,98 +8,71 @@ title: Spark on Kubernetes Integration Tests
-image tag that is appropriate.
-Here is an example:
-
-```
-$ mvn clean integration-test  \
-    -Dspark-distro-tgz=spark/spark-2.3.0-SNAPSHOT-bin.tgz  \
-    -Dspark.kubernetes.test.imageDockerTag=latest
-```
+The simplest way to run the integration tests is to install and run Minikube, then run the following:
+
+    build/mvn integration-test
+
+The minimum tested version of Minikube is 0.23.0. The kube-dns addon must be enabled. Minikube should
+run with a minimum of 3 CPUs and 4G of memory:
+
+    minikube start --cpus 3 --memory 4G
+
+You can download Minikube [here](https://github.com/kubernetes/minikube/releases).
+
+# Integration test customization
+
+Configuration of the integration test runtime is done through passing different Java system properties to the Maven
+command. The main useful options are outlined below.
+
+## Use a non-local cluster
+
+To use your own cluster running in the cloud, set the following:
+
+* `spark.kubernetes.test.deployMode` to `cloud` to indicate that Minikube will not be used.
+* `spark.kubernetes.test.master` to your cluster's externally accessible URL
+* `spark.kubernetes.test.imageRepo` to a write-accessible Docker image repository that provides the images for your
+cluster. The framework assumes your local Docker client can push to this repository.
+
+Therefore the command looks like this:
+
+    build/mvn integration-test \
+      -Dspark.kubernetes.test.deployMode=cloud \
+      -Dspark.kubernetes.test.master=https://example.com:8443/apiserver \
+      -Dspark.kubernetes.test.repo=docker.example.com/spark-images
+
+## Re-using Docker Images
+
+By default, the test framework will build new Docker images on every test execution. A unique image tag is generated,
+and it is written to file at `target/imageTag.txt`. To reuse the images built in a previous run, set:
+
+* `spark.kubernetes.test.imageTag` to the tag specified in `target/imageTag.txt`
+* `spark.kubernetes.test.skipBuildingImages` to `true`
+
+Therefore the command looks like this:
+
+    build/mvn integration-test \
+      -Dspark.kubernetes.test.imageTag=$(cat target/imageTag.txt) \
+      -Dspark.kubernetes.test.skipBuildingImages=true
+
+## Customizing the Spark Source Code to Test
+
+By default, the test framework will test the master branch of Spark from [here](https://github.com/apache/spark). You
+can specify the following options to test against different source versions of Spark:
+
+* `spark.kubernetes.test.sparkRepo` to the git or http URI of the Spark git repository to clone
+* `spark.kubernetes.test.sparkBranch` to the branch of the repository to build.
+
+An example:
+
+    build/mvn integration-test \
+      -Dspark.kubernetes.test.sparkRepo=https://github.com/apache-spark-on-k8s/spark \
+      -Dspark.kubernetes.test.sparkBranch=new-feature
+
+Additionally, you can use a pre-built Spark distribution. In this case, the repository is not cloned at all, and no
```

Comment:
I don't understand the motivation of why we should have the integration test code ever try to clone and build spark. Why not just always depend on a pre-built spark distribution? (Sorry if I missed the reason for this)

Reply:
I've found it a lot easier locally to be able to run a single command that both fetches the Spark distribution and builds it and then the integration test uses that Spark distribution. We provide the optionality for the local development scenario but we can still provide the TGZ specifically in Jenkins via `spark.kubernetes.test.sparkTgz`.